Testnet/MintableTokenInvoice/MintableTokenInvoice/MintableTokenInvoice.cs

```diff
@@ -91,61 +91,102 @@ public string GetInvoiceReference(string transactionReference)
         return $"INV-{invoiceReference.Substring(0, 4)}-{invoiceReference.Substring(4, 4)}-{invoiceReference.Substring(8, 4)}";
     }
 
-    private string ValidateKYC(Address sender, string invoiceReference)
+    private void EnsureKYCdUserOnly(Address sender)
     {
         // KYC check. Call Identity contract.
         ITransferResult result = this.Call(IdentityContract, 0, "GetClaim", new object[] { sender, KYCProvider });
-        if (!(result?.Success ?? false))
-        {
-            string reason = "Could not determine KYC status";
-            Log(new InvoiceResult() { InvoiceReference = invoiceReference, Success = false, Reason = reason });
-            return reason;
-        }
+        Assert(result.Success, "Could not determine KYC status");
 
-        // The return value is a json string encoding of a Model.Claim object, represented as a byte array using ascii encoding.
-        // The "Key" and "Description" fields of the json-encoded "Claim" object are expected to contain "Identity Approved".
-        if (result.ReturnValue == null || !Serializer.ToString((byte[])result.ReturnValue).Contains("Identity Approved"))
-        {
-            string reason = "Your KYC status is not valid";
-            Log(new InvoiceResult() { InvoiceReference = invoiceReference, Success = false, Reason = reason });
-            return reason;
-        }
-
-        return string.Empty;
+        // Only verified users are saved in the Identity contract.
+        Assert(result.ReturnValue != null && ((byte[])result.ReturnValue).Length != 0, "Your KYC status is not valid");
     }
 
-    /// <inheritdoc />
-    public string CreateInvoice(string symbol, UInt256 amount, UInt128 uniqueNumber)
+    private string CreateInvoiceInternal(Address address, string symbol, UInt256 amount, UInt128 uniqueNumber, string targetAddress, string targetNetwork)
```

Comment:
This method should be refactored according what we have talked. This method can succeed only once and all fail cases will not affect already state. 

Also how this method prevents duplicate call based on uid parameter ?

Reply:
The call is specifically allowed to be made multiple times as long as the arguments are not changed. The purpose is to allow: (1) KYC fails in the first call, perform KYC, retry call.

```
        if (invoice.To != Address.Zero)
            // If called with the same unique number then the details should not change.
            Assert(invoice.To == address && invoice.Symbol == symbol && invoice.Amount == amount && invoice.TargetNetwork == targetAddress && invoice.TargetNetwork == targetNetwork, "Transaction reference already exists");
       
```

Reply:
It supports the flow where e-mail is sent for "create invoice" link with embedded unique number. (1) User clicks link in e-mail to create invoice, (2) KYC fails, (3) User performs KYC, (4) User link in e-mail again to retry with same UID.

This is how we discussed with Iain and it has been recognized as a possible future improvement to the flow.

Reply:
The alternative is not to fail CreateInvoice for KYC or Approval checks and leave those checks exclusively to RetrieveInvoice. This may require some code updates to the daemons.

Reply:
Unfortunately the whole mechanism/approach is broken now due to introducing Asserts for KYC or Approval failures.